File: R/lehmann.R
power.ladesign <- function(gsize, odds.ratio, sig.level = 0.05, statistic = c("Kruskal-Wallis", "Jonckheere-Terpstra"), alternative = c("two.sided", "one.sided"), nrep=1e+6) {
  ng <- length(gsize)
  nOR <- length(odds.ratio)
  if (nOR != ng-1) {
    errmesg <- paste(" There are", ng, "groups and", nOR,"odds ratios, there should be", ng-1)
    stop(message=errmesg)
  }
  statistic <- match.arg(statistic)
  kw <- statistic=="Kruskal-Wallis"
  alternative <- match.arg(alternative)
  if (kw & alternative == "one.sided") stop("Only two-sided alternative is allowed for Kruskal-Wallis test")
  ng <- length(gsize)
  # number of possible combinations
  ncomb <- exp(lgamma(sum(gsize)+1) - sum(lgamma(gsize+1)))
  if (ncomb > 1e5) {
      zz <- .Fortran("lehman",
                     as.integer(ng),
                     as.integer(gsize),
                     double(ng),
                     as.double(rep(1, ng)),
                     as.double(sum(gsize)),
                     double(ng),
                     as.logical(kw),
                     as.integer(nrep),
                     tstat=double(nrep))
  } else {
      zz <- exactNull(gsize, kw)
  }
  if (kw) {
      # round the statistic to 3 significant digits after decimal
      ldq <- quantile(round(zz$tstat, 3), 1-sig.level)
  } else {
      if (alternative=="one.sided") {
          ldq <- quantile(zz$tstat, c(sig.level, 1-sig.level))
      } else {
          ldq <- quantile(zz$tstat, c(sig.level/2, 1-sig.level/2))
      }
  }
  zz <- .Fortran("lehman",
                 as.integer(ng),
                 as.integer(gsize),
                 double(ng),
                 as.double(c(1,odds.ratio)),
                 as.double(sum(gsize*c(1,odds.ratio))),
                 double(ng),
                 as.logical(kw),                 
                 as.integer(nrep),
                 tstat=double(nrep))
  if (kw) {
      # round the statistic to 3 significant digits after decimal
      ldpow <- sum(round(zz$tstat, 3) >= ldq)/nrep
  } else {
      if (alternative=="one.sided") {
          ldpow <- max(sum(zz$tstat <= ldq[1]), sum(zz$tstat >= ldq[2]))/nrep
      } else {
          ldpow <- (sum(zz$tstat <= ldq[1]) + sum(zz$tstat >= ldq[2]))/nrep
      }
  }
  out <- list()
  out$group.size <- gsize
  out$odds.ratio <- odds.ratio
  out$statistic <- statistic
  out$alternative <- alternative
  out$sig.level <- sig.level
  out$power <- ldpow
  class(out) <- "ladesign"
  out
}

print.ladesign <- function(x, ...) {
  cat("         Number of groups =", length(x$group.size), "\n")
  cat("               group size =", x$group.size, "\n")
  cat("odds ratios w.r.t group 1 =", round(x$odds.ratio, 3), "\n")
  cat("           test statistic =", x$statistic, "\n")
  cat("              alternative =", x$alternative, "\n")
  cat("       significance level =", x$sig.level, "\n")
  cat("                    power =", round(x$power, 3), "\n")
  invisible(x)
}

# k groups of sizes n_1,...,n_k; Total is N
# number of combinations is factorial(N)/product(factorial(n_i))
# if it is small we can enumerate the whole set
exactNull <- function(gsize, kw) {
    ng <- length(gsize)
    # use function combn to recursively generate combinations
    allcombs <- list()
    N <- Ni <- sum(gsize)
    ncomb <- 1
    for (i in 1:(ng-1)) {
        allcombs[[i]] <- combn(Ni, gsize[i])
        ncomb <- ncomb*ncol(allcombs[[i]])
        Ni <- Ni - gsize[i]
    }
    # vector of test statistics
    tstat <- rep(0, ncomb)
    # ways of assigning sum(gsize) elements into ng groups of size gsize
    igrps <- rep(0, N)
    # get the rank sum for each assignment
    # for KW the observations are ranked globally
    # for JT group j is ranked among observations from groups j thru ng
    if (kw) {rsums <- rep(0, ng)} else {rsums <- rep(0, ng-1)}
    # loop through the columns of each element in allcombs in a nested way
    ids <- rep(1, ng-1)
    nmax <- unlist(lapply(allcombs, ncol))
    irow <- 0
    while (all(ids <= nmax)) {
        # increment irow
        irow <- irow + 1
        # the column index for each group is specified by ids[i]
        igrps <- rep(0, N)
        for (i in 1:(ng-1)) {
            if (kw) {
                ii <- which(igrps==0)[allcombs[[i]][, ids[i]]]
                igrps[ii] <- i
                rsums[i] <- sum(ii)
            } else {
                rsums[i] <- sum(allcombs[[i]][, ids[i]])
            }
        }
        # igrps[which(igrps==0)] <- ng
        # rank sum of the last group for KW
        if (kw) {rsums[ng] <- sum(which(igrps==0))}
        # compute KW or JT statistic for each of the row of grps
        if (kw) {
            tstat[irow] <- sum(rsums^2/gsize)
        } else {
            tstat[irow] <- sum(rsums)
        }
        # increment ids to the next possible combination
        # find the last index of ids below nmax
        jj <- which(ids < nmax)
        if (length(jj) > 0) {
            jj0 <- max(jj)
            # increment the value at jj0 and set the ones after it to 1
            ids[jj0] <- ids[jj0] + 1
            if (jj0 < ng-1) ids[(jj0+1):(ng-1)] <- 1
        } else {
            ids[1] <- nmax[1] + 1
        }
    }
    # convert the JT statistic from (Wilcoxon) rank sum to Mann-Whitney form
    if (!kw) tstat <- tstat - sum(gsize[-ng]*(gsize[-ng]+1)/2)
    # return the statistics
    list(tstat=tstat)
}

File: src/lehmann.f
c     Mann-Whitney, Kruskal-Wallis, Jonckheere-Terpstra test
c     (monte carlo) power under Lehmann alternative

c     ng     = number of groups
c     gsize  = group sizes
c     mcgsiz = temporary group size vector for monte carlo step
c     oratio = odds ratio
c     gsor   = inner product of group size and odds ratio
c     rsum   = rank-sum vector
c     nrep   = number of replications (for monte carlo)
c     kw     = indicator of Kruskal-Wallis or Jonckheere-Terpstra
c     tstat  = test statistic vector

      subroutine lehman(ng, gsize, mcgsiz, oratio, gsor, rsum, kw, nrep,
     1     tstat)
      integer ng, gsize(ng), nrep
      logical kw
      double precision mcgsiz(ng), oratio(ng), gsor, rsum(ng), 
     1     tstat(nrep)

      double precision jtstat
      external jtstat

      integer nn, nr, i

c     calculate the total sample size
      nn = 0
      do 10 i = 1, ng
         nn = nn + gsize(i)
 10   continue

      call rndstart()
      if (kw) then
         do 50 nr = 1, nrep
c     initialize rank-sum and mcgsiz (= gsize*oratio) vector
            do 20 i = 1, ng
               rsum(i) = 0.0
               mcgsiz(i) = dfloat(gsize(i))*oratio(i)
 20         continue
c     generate a new rank-sum vector
            call kwrsum(nn, ng, mcgsiz, oratio, rsum, gsor)
c     compute Kruskal-Wallis test statistic
            tstat(nr) = 0.0
            do 30 i = 1, ng
               tstat(nr) = tstat(nr) + rsum(i)**2/dfloat(gsize(i))
 30         continue
 50      continue
      else
         do 100 nr = 1, nrep
c     initialize rank-sum and mcgsiz (= gsize*oratio) vector
            do 60 i = 1, ng
               rsum(i) = dfloat(gsize(i))
               mcgsiz(i) = dfloat(gsize(i))*oratio(i)
 60         continue
c     compute Jonckheere-Terpstra statistic for a new rank-sum vector
            tstat(nr) = jtstat(nn, ng, mcgsiz, oratio, rsum, gsor)
 100     continue
      endif
      call rndend()

      return
      end

      subroutine kwrsum(nn, ng, mcgsiz, oratio, rsum, gsor)
      integer nn, ng
      double precision mcgsiz(ng), oratio(ng), rsum(ng), gsor

      double precision dunif
      external dunif

      double precision cc, thrsh, mcgsor
      integer i, j

c     ith obsn is in group j with prob nj*or(j)/sum(nj*or(j))
c     sum(nj*or(j)) at the beginning 
      mcgsor = gsor
      do 20 i = 1, nn
c     scale the uniform random number to avoid division
         cc = mcgsor*dunif()
         j = 0
         thrsh = 0
c     check if the random number is within range for group j
         do 10 while(cc .gt. thrsh)
c     go the next group
            j = j + 1
c     increment threshold by nj*or(j)
            thrsh = thrsh + mcgsiz(j)
 10      enddo
c     decrease nj*or(j) for the jth group
         mcgsiz(j) = mcgsiz(j) - oratio(j)
c     decrease sum(nj*or(j))
         mcgsor = mcgsor - oratio(j)
c     add to rank sum for jth group
         rsum(j) = rsum(j) + dfloat(i)
 20   continue

      return
      end

      double precision function jtstat(nn, ng, mcgsiz, oratio, rsum, 
     1     gsor)
      integer nn, ng
      double precision mcgsiz(ng), oratio(ng), rsum(ng), gsor

      double precision dunif
      external dunif

      double precision cc, thrsh, mcgsor
      integer i, j

c     initialize J-T statistic as n+(n-1)+...+1
      jtstat = dfloat(nn*(nn+1))/2
c     ith obsn is in group j with prob nj*or(j)/sum(nj*or(j))
c     sum(nj*or(j)) at the beginning 
      mcgsor = gsor
      do 20 i = 1, nn
c     scale the uniform random number to avoid division
         cc = mcgsor*dunif()
         j = 0
         thrsh = 0
c     check if the random number is within range for group j
         do 10 while(cc .gt. thrsh)
c     go the next group
            j = j + 1
c     decrease J-T statistic by number of subjects in group j
            jtstat = jtstat - rsum(j)
c     increment threshold by nj*or(j)
            thrsh = thrsh + mcgsiz(j)
 10      enddo
c         call dblepr("jtstat",6,jtstat,1)
c         call intpr(" j",2,j,1)
c     decrease nj*or(j) for the jth group
         mcgsiz(j) = mcgsiz(j) - oratio(j)
c     decrease sum(nj*or(j))
         mcgsor = mcgsor - oratio(j)
c     reduce jth group size
         rsum(j) = rsum(j) - 1.0
 20   continue

      return
      end
